Sources/Apollo/LegacyCacheReadInterceptor.swift

```diff
@@ -27,6 +30,35 @@ public class LegacyCacheReadInterceptor: ApolloInterceptor {
                          completion: completion)
     case .query:
       switch request.cachePolicy {
+      case .fetchReturningCacheDataOnError:
+        chain.proceedAsync(
```

Comment:
One thing I worry about here is that this sort of messes up the idea that interceptors should be called in the order they're provided, but I suppose this isn't _that_ different from a retry. 

I wonder if there's a way to accomplish this where if an error comes in from the network stack and this is the cache policy, we catch that and then set up the retry with a `returnCacheDataDontFetch` policy. Then we can just proceed as we do with the `fetchIgnoringCacheData` etc here, and then when it comes again as `returnCacheDataDon'tFetch` we've already got handling for it. 

What do you think?

Reply:
One other thing to keep in mind is that there might be more people want to customize when it comes to error handling, and we need to consider GraphQL errors in addition to network errors.

Apollo Client web has a configurable [error policy](https://www.apollographql.com/docs/react/data/error-handling/) to specify how GraphQL errors (as opposed to network errors) are handled for example. We don't really take these into account at all right now. So we store incomplete data into the cache even if a response contains errors for example. That should probably be configurable. And it will also interact with `fetchReturningCacheDataOnError`, because returning data from the cache might be preferable to returning incomplete data.

I think having sensible policies for common situations is important, but ideally these would be built on composable primitives. There is a lot to consider here, especially once we have TTL added to the mix, so I don't think we can cover all cases with predefined policies.

Reply:
While the error policy is definitely something great to evolve towards, I think it's a little orthogonal to this capability.

What if instead this becomes an optional configuration on the `NetworkFetchInterceptor` where the developer can determine what do to on an error, and it kicks off a retry like @designatednerd suggested?

Reply:
well, in theory it would likely need to be an interceptor (possibly even an `additionalErrorInterceptor`), since for instance something that comes through with like a 400 error would not be caught by the `NetworkFetchInterceptor`, but by the `ResponseCodeInterceptor`.